File: soul-server/src/soul_server/service/file_manager.py
"""
FileManager - 첨부 파일 관리

첨부 파일 업로드, 검증, 정리를 담당합니다.
"""

import time
import shutil
import mimetypes
from pathlib import Path
from typing import Optional

import aiofiles

from soul_server.constants import MAX_ATTACHMENT_SIZE, DANGEROUS_EXTENSIONS


class AttachmentError(Exception):
    """첨부 파일 처리 오류"""
    pass


class FileManager:
    """
    첨부 파일 관리자

    역할:
    1. 첨부 파일 저장 (스레드별 격리)
    2. 파일 검증 (크기, 확장자)
    3. 스레드 첨부 파일 정리
    """

    def __init__(
        self,
        base_dir: Optional[str] = None,
        max_size: int = MAX_ATTACHMENT_SIZE
    ):
        """
        Args:
            base_dir: 첨부 파일 저장 기본 디렉토리. 미지정 시 config에서 읽음.
            max_size: 최대 파일 크기 (bytes)
        """
        if base_dir:
            self._base_dir = Path(base_dir)
        else:
            from soul_server.config import get_settings
            self._base_dir = Path(get_settings().incoming_file_dir)
        self._max_size = max_size
        self._base_dir.mkdir(parents=True, exist_ok=True)

    def get_thread_dir(self, thread_id: int) -> Path:
        """스레드별 첨부 파일 디렉토리"""
        thread_dir = self._base_dir / str(thread_id)
        thread_dir.mkdir(parents=True, exist_ok=True)
        return thread_dir

    def validate_file(self, filename: str, size: int) -> None:
        """
        파일 검증

        Args:
            filename: 파일명
            size: 파일 크기 (bytes)

        Raises:
            AttachmentError: 검증 실패
        """
        # 크기 검증
        if size > self._max_size:
            raise AttachmentError(
                f"파일이 너무 큽니다 ({size // 1024 // 1024}MB > {self._max_size // 1024 // 1024}MB)"
            )

        # 확장자 검증
        suffix = Path(filename).suffix.lower()
        if suffix in DANGEROUS_EXTENSIONS:
            raise AttachmentError(
                f"보안상 허용되지 않는 파일 형식입니다: {suffix}"
            )

    async def save_file(
        self,
        thread_id: int,
        filename: str,
        content: bytes
    ) -> dict:
        """
        파일 저장

        Args:
            thread_id: 스레드 ID
            filename: 원본 파일명
            content: 파일 내용

        Returns:
            저장 결과 딕셔너리

        Raises:
            AttachmentError: 검증 실패
        """
        # 검증
        self.validate_file(filename, len(content))

        # 파일명 안전하게 변환 (타임스탬프 추가)
        timestamp = int(time.time() * 1000)
        safe_filename = f"{timestamp}_{filename}"
        thread_dir = self.get_thread_dir(thread_id)
        file_path = thread_dir / safe_filename

        # 저장 (비동기 I/O)
        async with aiofiles.open(file_path, 'wb') as f:
            await f.write(content)

        # MIME 타입 추측
        content_type, _ = mimetypes.guess_type(filename)
        if not content_type:
            content_type = "application/octet-stream"

        return {
            "path": str(file_path),
            "filename": filename,
            "size": len(content),
            "content_type": content_type,
        }

    def is_safe_path(self, path: str, workspace_dir: Optional[str] = None) -> bool:
        """
        파일 경로 보안 검증

        Args:
            path: 검사할 경로
            workspace_dir: 허용된 워크스페이스 디렉토리

        Returns:
            True if safe, False otherwise
        """
        try:
            resolved = Path(path).resolve()
            resolved_str = str(resolved)

            # 허용된 디렉토리 확인
            allowed = False

            # 1. 첨부 파일 디렉토리
            if resolved_str.startswith(str(self._base_dir)):
                allowed = True

            # 2. 임시 Claude Code 디렉토리
            if resolved_str.startswith('/tmp/claude-code-'):
                allowed = True

            # 3. 워크스페이스 디렉토리
            if workspace_dir and resolved_str.startswith(workspace_dir):
                allowed = True

            if not allowed:
                return False

            # 위험한 확장자 체크
            if resolved.suffix.lower() in DANGEROUS_EXTENSIONS:
                return False

            # 파일 존재 확인
            if not resolved.exists():
                return False

            # 디렉토리 여부 확인
            if resolved.is_dir():
                return False

            # 크기 확인
            if resolved.stat().st_size > self._max_size:
                return False

            return True

        except Exception:
            return False

    def cleanup_thread(self, thread_id: int) -> int:
        """
        스레드의 첨부 파일 정리

        Args:
            thread_id: 스레드 ID

        Returns:
            삭제된 파일 수
        """
        thread_dir = self._base_dir / str(thread_id)
        if not thread_dir.exists():
            return 0

        try:
            # 파일 수 계산
            files_removed = sum(1 for _ in thread_dir.iterdir() if _.is_file())
            # 디렉토리 전체 삭제
            shutil.rmtree(thread_dir)
            return files_removed
        except Exception:
            return 0

    def cleanup_old_files(self, max_age_hours: int = 24) -> int:
        """
        오래된 첨부 파일 정리

        Args:
            max_age_hours: 최대 보관 시간 (시간)

        Returns:
            삭제된 디렉토리 수
        """
        max_age_seconds = max_age_hours * 3600
        current_time = time.time()
        cleaned = 0

        if not self._base_dir.exists():
            return 0

        for thread_dir in self._base_dir.iterdir():
            if not thread_dir.is_dir():
                continue

            try:
                # 디렉토리 수정 시간 확인
                mtime = thread_dir.stat().st_mtime
                if current_time - mtime > max_age_seconds:
                    shutil.rmtree(thread_dir)
                    cleaned += 1
            except Exception:
                continue

        return cleaned

    def get_stats(self) -> dict:
        """첨부 파일 통계"""
        total_files = 0
        total_size = 0
        thread_count = 0

        if self._base_dir.exists():
            for thread_dir in self._base_dir.iterdir():
                if thread_dir.is_dir():
                    thread_count += 1
                    for file_path in thread_dir.iterdir():
                        if file_path.is_file():
                            total_files += 1
                            total_size += file_path.stat().st_size

        return {
            "base_dir": str(self._base_dir),
            "thread_count": thread_count,
            "total_files": total_files,
            "total_size_mb": round(total_size / (1024 * 1024), 2),
            "max_file_size_mb": self._max_size // (1024 * 1024),
        }


# 싱글톤 인스턴스
file_manager = FileManager()

File: soul-server-ts/src/service/file_manager.ts
/**
 * FileManager — 첨부 파일 관리
 *
 * Python `soul-server/src/soul_server/service/file_manager.py` 동등 복제.
 * 디스크 구조: `{baseDir}/{sessionId}/{ts_ms}_{filename}`
 * 검증: 크기 ≤ MAX_ATTACHMENT_SIZE(8MB), 확장자 ∉ DANGEROUS_EXTENSIONS
 *
 * design-principles §1·§3·§4 정합:
 * - baseDir 미설정 시 throw (코드 default 없음)
 * - isUnderBase: realpath 기반 symlink 목적지 검증
 * - 정본은 이 클래스 하나
 */

import { mkdir, realpath, rm, stat, readFile, writeFile, readdir } from "node:fs/promises";
import * as path from "node:path";

import { MAX_ATTACHMENT_SIZE, DANGEROUS_EXTENSIONS } from "../constants.js";

/** Python `AttachmentError` 동등 */
export class AttachmentError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "AttachmentError";
  }
}

export interface SaveResult {
  /** 노드 절대경로 (Python `save_file_for_session` 반환 정합) */
  path: string;
  /** 타임스탬프 prefix가 붙은 저장 파일명 (`{ts_ms}_{filename}`) */
  filename: string;
  /** bytes */
  size: number;
  content_type: string;
}

export interface Stats {
  base_dir: string;
  /** 세션(하위 디렉토리) 수 */
  session_count: number;
  total_files: number;
  total_size_mb: number;
  max_file_size_mb: number;
}

/**
 * 파일 확장자 → MIME 타입 inline map (핵심 형식).
 * mime-types 패키지 없이 동작하도록 내장. Python `mimetypes.guess_type` 정합 fallback 포함.
 */
const MIME_MAP: Record<string, string> = {
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".webp": "image/webp",
  ".svg": "image/svg+xml",
  ".txt": "text/plain",
  ".md": "text/markdown",
  ".csv": "text/csv",
  ".json": "application/json",
  ".pdf": "application/pdf",
  ".zip": "application/zip",
  ".mp4": "video/mp4",
  ".mp3": "audio/mpeg",
  ".wav": "audio/wav",
  ".py": "text/x-python",
  ".ts": "text/typescript",
  ".js": "text/javascript",
  ".html": "text/html",
  ".css": "text/css",
  ".xml": "application/xml",
};

export function guessMimeType(filename: string): string {
  const ext = path.extname(filename).toLowerCase();
  return MIME_MAP[ext] ?? "application/octet-stream";
}

export class FileManager {
  private readonly baseDir: string;
  private readonly maxSize: number;

  /**
   * @param opts.baseDir - 첨부 파일 저장 기본 디렉토리. 미설정 시 throw.
   *   (env-variables.md §1: 코드 default 금지 — 호출자가 명시적으로 결정)
   * @param opts.maxSize - 최대 파일 크기 (bytes). default: MAX_ATTACHMENT_SIZE
   */
  constructor(opts: { baseDir?: string; maxSize?: number } = {}) {
    if (!opts.baseDir) {
      throw new Error(
        "FileManager: baseDir is required. Set INCOMING_FILE_DIR env or pass baseDir explicitly.",
      );
    }
    this.baseDir = opts.baseDir;
    this.maxSize = opts.maxSize ?? MAX_ATTACHMENT_SIZE;
  }

  /** baseDir 하위의 세션별 디렉토리 경로 (생성 포함) */
  async getSessionDir(sessionId: string): Promise<string> {
    const dir = path.join(this.baseDir, sessionId);
    await mkdir(dir, { recursive: true });
    return dir;
  }

  /**
   * absPath가 baseDir 하위에 위치하는지 검사 (directory traversal 방지).
   *
   * Python `is_under_base` 정합:
   * - fs.realpath로 symlink 목적지를 resolve한 뒤 base 하위인지 판정
   * - broken symlink, base 밖을 가리키는 symlink 둘 다 거부
   * - ValueError(relative_to 실패) / OSError(resolve 실패) 모두 false 반환
   */
  async isUnderBase(absPath: string): Promise<boolean> {
    try {
      const [resolvedTarget, resolvedBase] = await Promise.all([
        realpath(absPath),
        realpath(this.baseDir),
      ]);
      // path.relative가 '..'로 시작하면 base 하위 아님
      const rel = path.relative(resolvedBase, resolvedTarget);
      return !rel.startsWith("..") && !path.isAbsolute(rel);
    } catch {
      // ENOENT (broken symlink·존재하지 않는 경로), EACCES 등 — 보수적으로 거부
      return false;
    }
  }

  /**
   * 파일 검증.
   *
   * @throws {AttachmentError} 크기 초과 또는 위험 확장자
   */
  validateFile(filename: string, size: number): void {
    if (size > this.maxSize) {
      const sizeMb = Math.floor(size / 1024 / 1024);
      const maxMb = Math.floor(this.maxSize / 1024 / 1024);
      throw new AttachmentError(
        `파일이 너무 큽니다 (${sizeMb}MB > ${maxMb}MB)`,
      );
    }
    const ext = path.extname(filename).toLowerCase();
    if ((DANGEROUS_EXTENSIONS as readonly string[]).includes(ext)) {
      throw new AttachmentError(
        `보안상 허용되지 않는 파일 형식입니다: ${ext}`,
      );
    }
  }

  /**
   * 세션 생성 전 파일 업로드.
   *
   * 디스크 구조: `{baseDir}/{sessionId}/{ts_ms}_{filename}`
   * Python `save_file_for_session` (L207-251) 정합.
   *
   * @returns SaveResult — path는 노드 절대경로
   * @throws {AttachmentError} 검증 실패
   */
  async saveFileForSession(
    filename: string,
    content: Buffer,
    sessionId: string,
  ): Promise<SaveResult> {
    this.validateFile(filename, content.length);

    const sessionDir = await this.getSessionDir(sessionId);
    const timestamp = Date.now();
    const safeName = `${timestamp}_${filename}`;
    const filePath = path.join(sessionDir, safeName);

    await writeFile(filePath, content);

    return {
      path: path.resolve(filePath),
      filename: safeName,
      size: content.length,
      content_type: guessMimeType(filename),
    };
  }

  /**
   * 세션 첨부 파일 정리.
   *
   * Python `cleanup_session` (L253-273) 정합.
   *
   * @returns 삭제된 파일 수 (디렉토리 미존재 시 0)
   */
  async cleanupSession(sessionId: string): Promise<number> {
    const sessionDir = path.join(this.baseDir, sessionId);
    let filesRemoved = 0;
    try {
      const entries = await readdir(sessionDir);
      for (const entry of entries) {
        try {
          const s = await stat(path.join(sessionDir, entry));
          if (s.isFile()) filesRemoved++;
        } catch {
          // stat 실패 — 카운트만 skip
        }
      }
      await rm(sessionDir, { recursive: true, force: true });
    } catch (err: unknown) {
      // ENOENT: 디렉토리 미존재 → 0 반환
      const code = (err as NodeJS.ErrnoException).code;
      if (code === "ENOENT") return 0;
      // 기타 오류도 보수적으로 0 반환 (Python 정합)
      return 0;
    }
    return filesRemoved;
  }

  /**
   * 오래된 첨부 파일 정리.
   *
   * Python `cleanup_old_files` (L298-328) 정합.
   * baseDir 바로 아래 디렉토리(세션 디렉토리)를 순회하며 mtime 기준으로 제거.
   *
   * @param maxAgeHours - 최대 보관 시간 (hours). default: 24
   * @returns 삭제된 디렉토리 수
   */
  async cleanupOldFiles(maxAgeHours = 24): Promise<number> {
    const maxAgeMs = maxAgeHours * 3600 * 1000;
    const now = Date.now();
    let cleaned = 0;

    let entries: string[];
    try {
      entries = await readdir(this.baseDir);
    } catch {
      return 0;
    }

    for (const entry of entries) {
      const dirPath = path.join(this.baseDir, entry);
      try {
        const s = await stat(dirPath);
        if (!s.isDirectory()) continue;
        if (now - s.mtimeMs > maxAgeMs) {
          await rm(dirPath, { recursive: true, force: true });
          cleaned++;
        }
      } catch {
        continue;
      }
    }
    return cleaned;
  }

  /** 첨부 파일 통계 (Python `get_stats` L330-351 정합) */
  async getStats(): Promise<Stats> {
    let totalFiles = 0;
    let totalSize = 0;
    let sessionCount = 0;

    let entries: string[] = [];
    try {
      entries = await readdir(this.baseDir);
    } catch {
      // baseDir 미존재 등
    }

    for (const entry of entries) {
      const dirPath = path.join(this.baseDir, entry);
      try {
        const s = await stat(dirPath);
        if (!s.isDirectory()) continue;
        sessionCount++;
        const files = await readdir(dirPath);
        for (const file of files) {
          try {
            const fs = await stat(path.join(dirPath, file));
            if (fs.isFile()) {
              totalFiles++;
              totalSize += fs.size;
            }
          } catch {
            // skip
          }
        }
      } catch {
        continue;
      }
    }

    return {
      base_dir: this.baseDir,
      session_count: sessionCount,
      total_files: totalFiles,
      total_size_mb: Math.round((totalSize / (1024 * 1024)) * 100) / 100,
      max_file_size_mb: Math.floor(this.maxSize / (1024 * 1024)),
    };
  }
}
